File: Cpp/mathutil.cpp
// mathutil.cpp


#include <iostream>
#include <algorithm>
#include <cmath>
#include <array>
#include <vector>
#include <utility>
#include <stdexcept>
#include "mathutil.h"
using namespace std;


/*
 * arithmetic utility
 */


int64_t Mathutil::sum(int64_t from, int64_t to, int64_t diff) {
    if (to < from) return 0;

    int64_t size = (to - from) / diff + 1;
    to = from + (size - 1) * diff;

    if ((from + to) % 2 == 0)
        return (from + to) / 2 * size;
    return size / 2 * (from + to);
}


size_t Mathutil::pow(size_t base, size_t exp) {
    size_t result = 1;
    while (exp) {
        if (exp & 1) result *= base;
        exp >>= 1;
        result *= result;
    }
    return result;

    /*
    if (exp == 0)
        return 1;

    if (exp % 2 == 0)
        return pow(base*base, exp/2);

    else
        return pow(base*base, exp/2) * base;
    */
}


size_t Mathutil::pow(size_t base, size_t exp, size_t mod) {
    size_t result = 1;
    base %= mod;
    while (exp) {
        if (exp & 1) result = result * base % mod;
        exp >>= 1;
        result = result * result % mod;
    }
    return result;
}


size_t pow_mod(size_t base, size_t exp, size_t mod) {
    size_t result = 1;
    base %= mod;
    for (size_t i = 0; i < exp; ++i)
        result = result * base % mod;

    return result;
}


// less possible to overflow
// revert to
size_t Mathutil::pow_s(size_t base, size_t exp, size_t mod) {
    size_t result = 1;
    base %= mod;
    while (exp) {
        if (exp & 1) {
            if (result > numeric_limits<size_t>::max() / base) {
                // throw overflow_error("overflow in computing modular exponentiation");
                cout << "overflow for exponent algorithm, using product algorithm" << endl;
                return pow_mod(base, exp, mod);
            }
            result = result * base % mod;
        }

        exp >>= 1;

        if (result > numeric_limits<size_t>::max() / result) {
            // throw overflow_error("overflow in computing modular exponentiation");
            return pow_mod(base, exp, mod);
        }
        result = result * result % mod;
    }
    return result;
}


// https://en.wikipedia.org/wiki/Binary_GCD_algorithm
int64_t Mathutil::gcd(int64_t u, int64_t v) {
    int shift;

    if (u == 0) return v;
    if (v == 0) return u;

    for (shift = 0; ((u | v) & 1) == 0; ++shift) {
        u >>= 1;
        v >>= 1;
    }

    while ((u & 1) == 0)
        u >>= 1;

    // from here on, u is always odd
    do {
        while ((v & 1) == 0)
            v >>= 1;

        if (u > v)
            swap(u, v);

        v -= u;
    } while (v != 0);

    return u << shift;
}


/*
 * combinatoric
 */


// possible overflow
int64_t Mathutil::nCr(int n, int k) {
    if (n < k) return 0;

    if (k > n / 2) k = n - k;

    int64_t result = 1;
    for (int i = 1; i <= k; ++i) {
        result *= n--;
        result /= i;
    }
    return result;
}


// overflow safe
uint64_t Mathutil::nCr_s(unsigned int n, unsigned int k) {
    if (k > n) return 0;

    if (k > n / 2) k = n - k;

    uint64_t result = 1;
    for (unsigned int i = 1; i <= k; ++i) {
        uint64_t g = gcd(result, i);
        result /= g;
        uint64_t t = n-- / (i / g);

        if (result > numeric_limits<uint64_t>::max() / t) {
            throw overflow_error("overflow in computing combination");
        }

        result *= t;
    }

    return result;
}


/*
 * dealing with prime numbers
 */


bool Mathutil::is_prime(const size_t n) {
    if (n <= 3) return n >= 2;
    if (!(n & 1) || n % 3 == 0) return false;

    for (size_t p = 5, limit = (size_t)sqrt(n)+1; p < limit; p += 6) {
        if (n % p == 0) return false;
        if (n % (p+2) == 0) return false;
    }
    return true;
}


/*
 * n-1 = 2^s * d
 *
 * a^d % n == 1  OR
 * a^(2^r*d) % n == -1   0 <= r <= s-1
 */

bool witness(size_t /*n*/, size_t /*a*/, size_t /*s*/, size_t /*d*/) {
    // size_t rem = Mathutil::pow_s(a, d, n);
    // if (rem == 1) return true;  // probable prime

    // for (size_t i = 0; i < s; ++i) {
    //     rem = rem * rem % n;

    //     if (rem == 1)   return false;
    //     if (rem == n-1) return true;  // probable prime
    // }

    return false;
}

/*
 * if n < 1,373,653, it is enough to test a = 2 and 3;
 * if n < 9,080,191, it is enough to test a = 31 and 73;
 * if n < 4,759,123,141, it is enough to test a = 2, 7, and 61;
 * if n < 1,122,004,669,633, it is enough to test a = 2, 13, 23, and 1662803;
 * if n < 2,152,302,898,747, it is enough to test a = 2, 3, 5, 7, and 11;
 * if n < 3,474,749,660,383, it is enough to test a = 2, 3, 5, 7, 11, and 13;
 * if n < 341,550,071,728,321, it is enough to test a = 2, 3, 5, 7, 11, 13, and 17.
 *
 * if n < 18,446,744,073,709,551,616 = 2^64,
 * it is enough to test a = 2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, and 37
 */

bool Mathutil::miller_rabin(size_t n) {
    if (n <= 3) return n >= 2;
    if (!(n & 1) || n % 3 == 0) return false;

    size_t s = 0;
    size_t d = n-1;

    while (!(d & 1)) {
        d /= 2;
        ++s;
    }

    array<size_t, 12> small_primes {{
            2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37}};

    for (auto p : small_primes) {
        if (!witness(n, p, s, d))
            return false;
    }

    return true;
}


/*
 * @param
 * primes: empty integer vector with enough space resevered (preferred)
 */
size_t Mathutil::prime_under(vector<int> &primes, size_t n) {
    size_t size = 0;
    vector<bool> isPrime(n, true);

    for (unsigned int prime = 2; prime < n; ++prime) {
        if (isPrime[prime]) {
            primes.push_back(prime);
            ++size;

            for (uint64_t next = (uint64_t)prime * prime; next < n; next += prime)
                isPrime[next] = false;
        }
    }

    return size;
}

File: Java/MathUtil.java
import java.math.BigInteger;
import java.util.*;

/**
 * Created by yxy on 3/12/2015.
 */
public class MathUtil {
    public static long pow(long base, long exp) {
        long result = 1;
        long pow = base;

        while (exp > 0) {
            if ((exp & 1) == 1) result *= pow;

            exp >>= 1;
            pow *= pow;
        }

        return result;
    }

    public static long pow(long base, long exp, long mod) {
        long power = 1;

        while (exp > 0) {
            if ((exp & 1) == 1) power = power * base % mod;

            exp >>= 1;
            base = base * base % mod;
        }

        return power;
    }

    /**
     safer version of pow, with less probability to overflow
     */
    public static long powMod(long base, long exp, long mod) {
        long result = 1;
        base %= mod;

        for (int i = 0; i < exp; i++) {
            result = result * base % mod;
        }

        return result;
    }


    private static final BigInteger TWO = BigInteger.valueOf(2);

    /**
     * Computes the integer square root of a number.
     *
     * @param n  The number.
     *
     * @return  The integer square root, i.e. the largest number whose square
     *     doesn't exceed n.
     */
    public static BigInteger sqrt(BigInteger n)
    {
        if (n.signum() >= 0)
        {
            final int bitLength = n.bitLength();
            BigInteger root = BigInteger.ONE.shiftLeft(bitLength / 2);

            while (!isSqrt(n, root))
            {
                root = root.add(n.divide(root)).divide(TWO);
            }
            return root;
        }
        else
        {
            throw new ArithmeticException("square root of negative number");
        }
    }

    private static boolean isSqrt(BigInteger n, BigInteger root)
    {
        final BigInteger lowerBound = root.pow(2);
        final BigInteger upperBound = root.add(BigInteger.ONE).pow(2);
        return lowerBound.compareTo(n) <= 0
                && n.compareTo(upperBound) < 0;
    }


    public static boolean isSquare(long n) {
        long root = Math.round(Math.sqrt(n));
        return root * root == n;
    }

    public static boolean isSquare(BigInteger n) {
        BigInteger root = sqrt(n);
        return root.multiply(root).equals(n);
    }

    public static int sum(List<Integer> integers) {
        int result = 0;
        for (int i : integers) {
            result += i;
        }
        return result;
    }

    public static int product(List<Integer> integers) {
        int result = 1;
        for (int i : integers) {
            result *= i;
        }
        return result;
    }

    public static int gcd(int x, int y) {
        while (y != 0) {
            int tmp = x % y;
            x = y;
            y = tmp;
        }
        return x;
    }

    public static long gcd(long x, long y) {
        while (y != 0) {
            long tmp = x % y;
            x = y;
            y = tmp;
        }
        return x;
    }

    public static long nCr(long n, long r) {
        if (r > n / 2) r = n - r;

        long result = 1;
        for (int i = 1; i <= r; i++) {
            result *= n - i + 1;
            result /= i;
        }

        return result;
    }

    public static BigInteger nCrBig(long n, long r) {
        if (r > n / 2) r = n - r;

        BigInteger result = BigInteger.ONE;
        for (int i = 1; i <= r; i++) {
            result = result.multiply(BigInteger.valueOf(n - i + 1));
            result = result.divide(BigInteger.valueOf(i));
        }

        return result;
    }

    /*
    public static BigInteger nCrBigMod(long n, long r, long mod) {
        if (r > n / 2) r = n - r;
        BigInteger modBig = BigInteger.valueOf(mod);

        BigInteger result = BigInteger.ONE;
        HashSet<Long> toDivide = new HashSet<Long>();
        for (long i = 1; i <= r; i++) {
            toDivide.add(i);
        }

        for (long i = n-r+1; i <= n; i++) {
            long current = i;

            Iterator<Long> iterator = toDivide.iterator();
            while (iterator.hasNext()) {
                Long element = iterator.next();
                if (current % element == 0) {
                    iterator.remove();
                    current /= element;
                }
            }

            result = result.multiply(BigInteger.valueOf(current)).mod(modBig);
        }

        return result;
    }*/

    public static boolean isPrime(int x) {
        if (x <= 3) return x >= 2;
        if (x % 2 == 0 || x % 3 == 0) return false;

        for (int i = 5, j = (int)Math.sqrt(x); i <= j; i += 6) {
            if (x % i == 0 || x % (i+2) == 0)
                return false;
        }

        return true;
    }

    public static boolean isPrimeLong(long x) {
        if (x <= 3) return x >= 2;
        if (x % 2 == 0 || x % 3 == 0) return false;

        for (long i = 5, j = (long)Math.sqrt(x); i <= j; i += 6) {
            if (x % i == 0 || x % (i+2) == 0)
                return false;
        }

        return true;
    }

    public static List<Integer> getPrime(int ceiling) {
        List<Integer> primes = new ArrayList<Integer>();
        boolean[] isPrime = new boolean[ceiling+1];

        for (int i = 0; i <= ceiling; i++)
            isPrime[i] = true;

        for (int p = 2; p <= ceiling; p++) {
            if (isPrime[p]) {
                primes.add(p);
                for (long i = (long)p*p; i <= ceiling; i += p) {
                    isPrime[(int)i] = false;
                }
            }
        }

        return primes;
    }

    public static BitSet isPrimeArray(int ceiling) {
        BitSet isPrime = new BitSet(ceiling);

        for (int i = 0; i <= ceiling; i++)
            isPrime.set(i);
        isPrime.set(0, false);
        isPrime.set(1, false);

        for (int p = 2; p <= ceiling; p++) {
            if (isPrime.get(p)) {
                for (long i = (long)p*p; i < ceiling; i += p) {
                    isPrime.set((int)i, false);
                }
            }
        }

        return isPrime;
    }

/*
    public static int[] getPrime(int lowerBound, int upperBound) {

    }*/

    public static boolean isPermutationOf(int x, int y) {
        char[] s1 = Integer.toString(x).toCharArray();
        char[] s2 = Integer.toString(y).toCharArray();
        if (s1.length != s2.length)
            return false;
        Arrays.sort(s1);
        Arrays.sort(s2);
        return Arrays.equals(s1, s2);
    }

    public static HashMap<Integer, Integer> primeFactorization(int x) {
        HashMap<Integer, Integer> result = new HashMap<Integer, Integer>();
        if (x % 2 == 0) {
            int exp = 1;
            x /= 2;
            while (x % 2 == 0) {
                exp ++;
                x /= 2;
            }
            result.put(2, exp);
        }

        int factor = 3;
        while (factor * factor <= x) {
            if (x % factor == 0) {
                int exp = 1;
                x /= factor;
                while (x % factor == 0) {
                    exp ++;
                    x /= factor;
                }
                result.put(factor, exp);
            }

            factor += 2;
        }

        if (x != 1) result.put(x, 1);

        return result;
    }

    public static ArrayList<ArrayList<Integer>> getAllFactors(int ceiling) {
        ArrayList<ArrayList<Integer>> allFactors = new ArrayList<ArrayList<Integer>>();
        for (int i = 0; i <= ceiling; i++) {
            allFactors.add(new ArrayList<Integer>());
        }

        boolean[] isPrime = new boolean[ceiling+1];
        for (int i = 0; i <= ceiling; i++) {
            isPrime[i] = true;
        }
        for (int i = 2; i <= ceiling; i++) {
            if (isPrime[i]) {
                allFactors.get(i).add(i);
                for (int j = 2*i; j <= ceiling; j+=i) {
                    allFactors.get(j).add(i);
                    isPrime[j] = false;
                }
            }
        }

        return allFactors;
    }

    public static void testMathUtil() {
        System.out.println(Long.MAX_VALUE + " " + Long.MIN_VALUE);
        System.out.println(Integer.MAX_VALUE + " " + Integer.MIN_VALUE);

        System.out.println("pow(12, 12) = " + pow(12, 12));
        System.out.println("pow(112413, 143, 213) = " + pow(112413, 143, 213));
        System.out.println("powMod(2, 25025, 10000000000) = " + powMod(2, 25025, pow(10, 10)));
        System.out.println("gcd(1341, 523893) = " + gcd(1341, 523893));
        System.out.println("nCr(45, 30) = " + nCr(45, 30));
        System.out.println("nCrBig(100, 50) = " + nCrBig(100, 50));
        // System.out.println("nCrBigMod(1000, 300, 1000000) = " + nCrBigMod(1000, 300, 1000000));
        System.out.println("isPrime(997) = " + isPrime(997));
        System.out.println("primeFactor(36) = " + primeFactorization(36));
        System.out.println("primeFactor(120) = " + primeFactorization(120));
        System.out.println("isPermutationOf(87109, 79180) = " + isPermutationOf(87109, 79180));

        System.out.println("getPrime(100) = " + getPrime(100));
        int primeUnder = (int)pow(10, 8);
        long getPrimeTime = System.nanoTime();
        System.out.println("getPrime(" + primeUnder + ")");
        List<Integer> primes = getPrime(primeUnder);
        System.out.println("500500th prime " + primes.get(500500-1));
        System.out.println("7037th prime " + primes.get(7037-1));
        /*for (int i: primes)
            System.out.print(i + " ");*/
        System.out.println("number of primes is " + primes.size() + "\nthe last prime is " + primes.get(primes.size()-1));
        System.out.println("primes acquired by " + (System.nanoTime()-getPrimeTime) / 1000000000.0 + " seconds");
        // roughly 7 seconds for primeUnder = 10 ^ 8, number of primes: 5761455

        /*long getSlowPrimeTime = System.nanoTime();
        List<Integer> primesSlow = new ArrayList<Integer>();
        for (int i = 11; i < primeUnder; i+=10)
            if (MathUtil.isPrime(i))
                primesSlow.add(i);
        for (int i = 13; i < primeUnder; i+=10)
            if (MathUtil.isPrime(i))
                primesSlow.add(i);
        for (int i = 17; i < primeUnder; i+=10)
            if (MathUtil.isPrime(i))
                primesSlow.add(i);
        for (int i = 19; i < primeUnder; i+=10)
            if (MathUtil.isPrime(i))
                primesSlow.add(i);
        System.out.println(primesSlow.get(primesSlow.size()-1));
        System.out.println("primes acquired by " + (System.nanoTime()-getSlowPrimeTime) / 1000000000.0 + " seconds");*/

        int ceiling = 1000000;
        long getAllFactorsTime = System.nanoTime();
        System.out.println("getAllFactors(" + ceiling + ")");
        ArrayList<ArrayList<Integer>> factors = getAllFactors(ceiling);
        System.out.println(factors.get(1000000-1));
        System.out.println("All factors acquired by " + (System.nanoTime()-getAllFactorsTime) / 1000000000.0 + " seconds");
    }

    public static void main(String[] args) {
        testMathUtil();
    }
}
